### PR TITLE
Update business form to create subscriptions

### PR DESCRIPTION
**Pull request by matthewdylan**
#### What's this PR do? 
Adds stripe subscription creation to our /business membership form. Also uses our new stripe webhooks functionality to listen for subscription creation success events from business membership submission and logs recurring donations and opportunities in Salesforce respectively.

#### Why are we doing this? How does it help us?
This is the next step in a larger modernization project for the donations app that will give us a more reliable and less brittle system for handling recurring donations. This update modernizes the /business donation form in particular.

#### How should this be manually tested?

* Add new business form env vars
     * Add `STRIPE_PRODUCT_BIG_TEX` `STRIPE_PRODUCT_LONE_STAR` and `STRIPE_PRODUCT_HATS_OFF` to your env-docker file. [Internal note: See wiki **Keys** for details](https://texastribune.atlassian.net/wiki/spaces/TECH/pages/2273837059/Navigating+Salesforce#Keys)
* Visit the [business form](https://donations-testing.herokuapp.com/business) on staging
* Try giving donations for different tiers (i.e. "Big Tex", "Lone Star", etc.), at different intervals, with or without agreeing to cover fees using a [test card](https://stripe.com/docs/testing#cards)
* For each of these donations...
    * In our stripe dashboard, make sure you are in "Test Mode" (upper corner)
    * Choose the "Customers" tab and you should see the gift you made
        * Note that it will show the First Name/Last Name combo you provided, not the company name 
        * Screen Shot 2023-07-07 at 7 21 30 AM
    * Clicking through to the customer you added should show a screen listing a Business subscription
    * Screen Shot 2023-07-07 at 7 22 22 AM
    * Login to our salesforce sandbox (deets listed under "Salesforce Test (2023 refactor)")
    * Follow the instructions [here](https://texastribune.atlassian.net/wiki/spaces/TECH/pages/2273837059/Navigating+Salesforce) to ensure that the donation came through on the salesforce end as expected

#### How should this change be communicated to end users?
We'll let membership team know when this is deployed.

#### Are there any smells or added technical debt to note?
The handling of stripe products is admittedly messy. The hope is that when we refactor the frontend, we can incorporate some more stripe pieces and have the products/prices loaded with the rest of the page instead of being referenced/grabbed here in the midst of the post request.

#### What are the relevant tickets?
https://airtable.com/appyo1zuQd8f4hBVx/tbloNZu8GkM52NKFR/viwS1XPty68eK4Ett/rec4vLTin8wF7HlrP?blocks=hide

#### Have you done the following, if applicable:
***(optional: add explanation between parentheses)***

* [ ] Added automated tests? *( )*
* [ ] Tested manually on mobile? *( )*
* [ ] Checked BrowserStack? *( )*
* [ ] Checked for performance implications? *( )*
* [ ] Checked accessibility? *( )*
* [ ] Checked for security implications? *( )*
* [ ] Updated the documentation/wiki? *( )*

#### TODOs / next steps:

* [ ] *your TODO here*

### CODE REVIEW

**Review by ashley-hebler — COMMENTED** [review]
Test steps seem to work for me. From what I could tell, everything was coming through Salesforce and Stripe as expected. It'd be good to have Anna C. verify the Saleforce side.

**Some observations:**

_Accounts and Contacts_
In testing, I didn't really intend to have all of my submissions bubble up to one account, but Salesforce grouped them this way. I don't think that's a reflection of the code changes — more-so just how I was testing. 
Screenshot 2023-07-10 at 10 59 37 AM

_Fees_
I was curious why the production version of of the front-end fee calculation was the same as my local version for both /business and /circle. My understanding is that [this line](https://github.com/texastribune/donations/blob/b9cb07787508eb1f2d55823210adc7c32c58792d/server/static/js/src/connected-elements/PayFees.vue#L64) means that we're already displaying the stripe transaction fee (.22) AND the stripe subscription fee (.05) for both business and circle currently. All that really means is that current circle/business donors end up paying less than we are displaying in the checkbox until we convert them to subscription products, which doesn't seem bad. Just seems like a bit of a gotcha as we live in both worlds for a bit!

File: server/app.py [hunk 1]
```diff
@@ -1390,6 +1274,49 @@ def create_subscription(customer=None, form=None, quarantine=None):
...
+    return subscription
+
+
+def find_price(prices=[], period=None, pay_fees=False):
+    nickname = "fees" if pay_fees else None
+    interval = "month" if period == "monthly" else "year"
+    for price in prices:
+        if price["recurring"]["interval"] == interval \
+            and price["nickname"] == nickname:
+            return price
```
[the remark this answers is not in the document]
Comment (matthewdylan) [reply server/app.py:1317]: Yeah, it's true that we would. That would only happen if someone was futzing with the product entries in slack or the code itself, which isn't likely, but it's not impossible. Also, adding some error handling here just makes the function read nicer, I think, so good suggestion.

File: server/app.py [hunk 2]
```diff
@@ -836,9 +850,18 @@ def payout_paid(event):
...
     customer = stripe.Customer.retrieve(subscription["customer"])
     contact = get_contact(customer)
-    rdo = log_rdo(contact, subscription)
+    if donation_type == "business_membership":
+        account = get_account(customer)
+        rdo = log_rdo(account=account, subscription=subscription)
+        affiliation = Affiliation.get_or_create(
+            account=account, contact=contact, role="Business Member Donor"
+        )
+        send_email_new_business_membership(account=account, contact=contact)
```
[the remark this answers is not in the document]
Comment (matthewdylan) [reply server/app.py:862]: Yeah, I didn't change anything about the way this func worked in the previous version so it shouldn't be effected, but yes, that's a great point about including it in our post-launch list.

**Comment by SimmonsRitchie** [comment]
Thanks for putting this together, @matthewdylan! Testing it now. One small thing, and this might be a bit outside the scope of this PR, but form UI-wise I think it would be ideal if users did not have to add the protocol when inputting their web address. This kept catching me out. I don't think it's very intuitive to most users to include that.

image

It also might be nice if the page auto-scrolled towards the error message when there is an invalid input (or, if there are multiple invalid inputs, scrolling to one of them). I found that i kept hitting 'submit' and I couldn't quite understand why nothing was happening. It wasn't until I scrolled up the page that I saw the error message. It's a small UI thing that I think can create a little more friction than ideal.

**Comment by SimmonsRitchie** [comment]
Hmmm, testing-wise, @matthewdylan, I made one Big Tex recurring donation that seemed to be successful. I see the details in both Stripe and Salesforce. However, I just made another recurring donation under a different name and I'm not sure if it's been successful.

I see it [here in Stripe](https://dashboard.stripe.com/test/customers/cus_OFEsCPxRxqRgzm) and I see a contact has been [created in Salesforce](https://texastribune--2023stripe.sandbox.lightning.force.com/lightning/r/Account/0016s00000cupwtAAA/view), but I don't see an opportunity related to this account? Am I just navigating the UI poorly or did something go awry here?

File: server/app.py [hunk 3]
```diff
@@ -1157,146 +1180,6 @@ def add_recurring_donation(contact=None, form=None, customer=None, quarantine=Fa
...
-    bad_actor_request=None,
-):
-    """
-    Adds a business membership. Both single and recurring.
-
-    It will look for a matching Contact (or create one). Then it will look for a
-    matching Account (or create one). Then it will add the single or recurring donation
-    to the Account. Then it will add an Affiliation to link the Contact with the
-    Account. It sends a notification to Slack (if configured). It will send email
-    notification about the new membership.
```
[the remark this answers is not in the document]
Comment (matthewdylan) [reply server/app.py:1207]: Good point. I can add this.

**Comment by matthewdylan** [comment]
> Test steps seem to work for me. From what I could tell, everything was coming through Salesforce and Stripe as expected. It'd be good to have Anna C. verify the Saleforce side.
> 
> **Some observations:**
> 
> _Accounts and Contacts_ In testing, I didn't really intend to have all of my submissions bubble up to one account, but Salesforce grouped them this way. I don't think that's a reflection of the code changes — more-so just how I was testing. Screenshot 2023-07-10 at 10 59 37 AM
> 
> _Fees_ I was curious why the production version of of the front-end fee calculation was the same as my local version for both /business and /circle. My understanding is that [this line](https://github.com/texastribune/donations/blob/b9cb07787508eb1f2d55823210adc7c32c58792d/server/static/js/src/connected-elements/PayFees.vue#L64) means that we're already displaying the stripe transaction fee (.22) AND the stripe subscription fee (.05) for both business and circle currently. All that really means is that current circle/business donors end up paying less than we are displaying in the checkbox until we convert them to subscription products, which doesn't seem bad. Just seems like a bit of a gotcha as we live in both worlds for a bit!

I wonder if all of your submissions bubbled up to one account because you gave the same website for each donation (there might be other linking factors like email, but I'm not certain).

On the transaction fee issue, good catch. That was an oversight on my part. I'd say we should just leave it as is for the /business page since we'll be deploying it relatively soon, but I wonder if it's worth trying to differentiate the /circle page's fee amount.

**Comment by matthewdylan** [comment]
> Hmmm, testing-wise, @matthewdylan, I made one Big Tex recurring donation that seemed to be successful. I see the details in both Stripe and Salesforce. However, I just made another recurring donation under a different name, Señor Crazy Cow, and I'm not sure if it's been successful.
> 
> I see it [here in Stripe](https://dashboard.stripe.com/test/customers/cus_OFEsCPxRxqRgzm) and I see a contact for Crazy Cow has been [created in Salesforce](https://texastribune--2023stripe.sandbox.lightning.force.com/lightning/r/Account/0016s00000cupwtAAA/view), but I don't see an "opportunity" related to Crazy Cow? Am I just navigating the UI poorly or did something go awry here?
> 
> image
> Looking a little closer, it looks like maybe Crazy Cow's donation is associated with [Testing Inc](https://texastribune--2023stripe.sandbox.lightning.force.com/lightning/r/Account/0016s00000cuTeEAAU/view)?
> 
> image
> I'm not quite sure why this would be the case, that wasn't Crazy Cow's company name (that would be La Vaca Loca Inc), but perhaps it grouped them because Testing Inc used the same card? I believe I used a JCB card for that transaction:
> 
> image
> Here's my original form Input (excluding card details):
> 
> image

I think the linking factor here is the website. I gave the first gift as Testing Inc. and I also used texastribune.org... did some digging and you can see how this works with Account.get_or_create() [here](https://github.com/texastribune/donations/blob/master/server/npsp.py#L893-L921).

File: server/app.py [hunk 4]
```diff
@@ -1390,6 +1274,49 @@ def create_subscription(customer=None, form=None, quarantine=None):
     return subscription
 
 
+# TODO can these funcs be moved somewhere else? (maybe util.py?)
+def create_subscription(donation_type=None, customer=None, form=None, quarantine=None):
+    app.logger.info(f"business form: {form}")
+    logging.info(f"business form: {form}")
```
[the remark this answers is not in the document]
Comment (ashley-hebler) [reply server/app.py:1280]: Make sense, thanks for clarifying and good note to have in our paper trail for why we have two separate functions

File: server/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashley-hebler) [reply server/app.py:862]: Sounds good! I added us to that email group for now so we can have some awareness around when that form is submitted. That might be nice to have in general anyway

**Pull request merged**